---
title: "Class 11: Structural Bioinformatics pt2."
author: "Osirus Eisenman"
format: pdf
editor: visual
---

## Background

We saw last day that the PDB has 209,886 entries (Oct/Nov 2025).UniProtKB (i.e. protein sequence database) has 199,579,901 entries

```{r}
209886/199579901*100

```

So the PDB has only 0.1% coverage of the main sequence database.

Enter AlphaFold data base (AFDB) \<<https://alphafold.ebi.ac.uk/>\> that attempts to provide computed models for all sequences in UniProt.

"AlphaFold structure predictions for over 200 million proteins are now available in the EBI AlphaFold Database (AFDB) (Figure 6). This is a result of an ongoing a collaboration between DeepMind and the European Bioinformatics Institute (EBI). The online interface allows for searching based on protein names, sequences, or accession numbers. Additionally, the database offers basic visualization of structure models with Mol\* and downloadable PDB files (Figure 7). If your protein of interest is there then you donâ€™t need to proceed with the timely step of running AlphaFold yourself (next section)."

## AlphaFold

AlphaFold has 3 main outputs

-the predicted coordinates (PDB files)

-A local quality score called **pLDDT** (one for each amino-acid)

-A second quality score **PAE** (Predicted Aligned Error) for each pair of amino-acids

We can run AlphaFold ourselves if we are not happy with AFDB (i.e. no coverage or poor model).

## Interpreting/ Analyzing AF Results in R

```{r}
# Change this for YOUR results dir name
results_dir <- "HIVPR_dimer.a4e10" 
```

```{r}
# File names for all PDB models
pdb_files <- list.files(path=results_dir,
                        pattern="*.pdb",
                        full.names = TRUE)

# Print our PDB file names
basename(pdb_files)

```

```{r}
library(bio3d)
library(bio3dview)

# Read all data from Models 
#  and superpose/fit coords
pdbs <- pdbaln(pdb_files, fit=TRUE, exefile="msa")
```

```{r}
rd <- rmsd(pdbs, fit=T)
```

```{r}
range(rd)
```

```{r}
library(pheatmap)

colnames(rd) <- paste0("m",1:5)
rownames(rd) <- paste0("m",1:5)
pheatmap(rd)
```

```{r}
# Read a reference PDB structure
pdb <- read.pdb("1hsg")
```

```{r}
plotb3(pdbs$b[1,], typ="l", lwd=2, sse=pdb)
points(pdbs$b[2,], typ="l", col="red")
points(pdbs$b[3,], typ="l", col="blue")
points(pdbs$b[4,], typ="l", col="darkgreen")
points(pdbs$b[5,], typ="l", col="orange")
abline(v=100, col="gray")
```
