---
title: "class13"
author: "Osirus Eisenman"
format: pdf
toc: TRUE
---

## Background:

Today we will run through a complete RNASeq analysis in this hands-on session comes from GEO entry: [GSE37704](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE37704), which is associated with the following publication:

-   Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "*Differential analysis of gene regulation at transcript resolution with RNA-seq*". Nat Biotechnol 2013 Jan;31(1):46-53. [PMID: 23222703](https://www.ncbi.nlm.nih.gov/pubmed/23222703)

The authors report on differential analysis of lung fibroblasts in response to loss of the developmental transcription factor HOXA1. Their results and others indicate that HOXA1 is required for lung fibroblast and HeLa cell cycle progression. In particular their analysis show that "loss of HOXA1 results in significant expression level changes in thousands of individual transcripts, along with isoform switching events in key regulators of the cell cycle". For our session we have used their [Sailfish](https://www.nature.com/articles/nbt.2862) gene-level estimated counts and hence are restricted to protein-coding genes only.

## Data Import:

```{r}
metadata<-read.csv("GSE37704_metadata.csv")
counts <- read.csv("featurecounts.csv")
```

Check correspondence of `metadata` and `counts` (i.e. that the columns in `counts` match the rows in the `metadata`).

```{r}
metadata
```

```{r}
colnames(counts)
```

```{r}
metadata$id
```

Fix to remove that first "length" column of `counts`

```{r}
counts<-counts[,-1]
```

Also let's remove low count genes

```{r}
tot.counts<-rowSums(counts)
head(tot.counts)
```

Let's remove all zero count genes

```{r}
zero.inds<-tot.counts==0
head(zero.inds)
```

```{r}
counts<-counts[!zero.inds,]
```

```{r}
test_cols<-!all(colnames(counts)==metadata$id)
```

```{r}
if (test_cols){
  message("Wow...these is a problem with the metadata counts setup")
}
```

## Setup for DESeq:

```{r}
library(DESeq2)
```

```{r}
dds<-DESeqDataSetFromMatrix(countData=counts,
                            colData=metadata,
                            design=~condition)
```

## Run DESeq:

```{r}
dds<-DESeq(dds)
```

## Get Results:

```{r}
res<-results(dds)
```

```{r}
head(res)
```

## Add Annotation:

```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)

columns(org.Hs.eg.db)

res$symbol <- mapIds(org.Hs.eg.db,
                     keys = res$ensembl,   
                     keytype = "ENSEMBL",
                     column = "SYMBOL",  
                     multiVals = "first")
#res$symbol = mapIds(org.Hs.eg.db,
                 #   keys=___, 
                  #  keytype="ENSEMBL",
                  #  column=___,
                   # multiVals="first")

res$entrez <- mapIds(org.Hs.eg.db,
                     keys = res$ensembl,  
                     keytype = "ENSEMBL",
                     column = "ENTREZID", 
                     multiVals = "first")

#res$entrez = mapIds(org.Hs.eg.db,
                  #  keys=___,
                   # keytype="ENSEMBL",
                   # column="ENTREZID",
                   # multiVals="first")

res$name <- mapIds(org.Hs.eg.db,
                   keys = row.names(res),   
                   keytype = "ENSEMBL",
                   column = "GENENAME",   
                   multiVals = "first")

#res$name =   mapIds(org.Hs.eg.db,
                   # keys=row.names(res),
                  #  keytype=___,
                  #  column=___,
                   # multiVals="first")

head(res, 10)
```

## Visualize Results:

```{r}
library(ggplot2)
```

```{r}
my_cols<-rep("gray",nrow(res))
my_cols[abs(res$log2FoldChange)>=2]<-"blue"


ggplot(res)+
  aes(log2FoldChange,-log(padj))+
  geom_point(col=my_cols)+
  geom_hline(yintercept = -log(0.05),col="red")+
  geom_vline(xintercept = c(-2,2),col="red")
```

## Pathway Analysis:

#BiocManager::install( c("pathview", "gage", "gageData") )

```{r}
library(gage)
library(gageData)

data(kegg.sets.hs)
data(sigmet.idx.hs)
```

For **gage** we want a named vector of importance

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

```{r}
keggres=gage(foldchanges,gsets=kegg.sets.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]
```

```{r}
head(keggres$less)
head(kegg.sets.hs, 3)
```

```{r,message=FALSE}
library(pathview)
pathview()
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```

```{r}
# A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
```

```{r}
# A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
```

```{r}
## Focus on top 5 upregulated pathways here for demo purposes only
keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```

```{r}
pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")
```

!{}(.png)

## GO Analysis:

```{r}
data(go.sets.hs) data(go.subs.hs)  # Focus on Biological Process subset of GO 
gobpsets <- go.sets.hs[go.subs.hs$BP]  
gobpres <- gage(foldchanges, gsets=gobpsets, same.dir=TRUE) 
```

```{r}
head(gobpres$less)
lapply(gobpres, head)
```

## Reactome:

Some folks really like Reactome online (i.e. their webpage viewer (<https://reactome.org/>)) rather than the R package of the same name (available from bioconductor).

To use the website viewer we want to upload our set of gene symbols for the genes we want to focus on (here those with a p-value below 0.05).

```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]

write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)

print(paste("Total number of significant genes:", length(sig_genes)))
```

\![\](.png)

## Save Results:

```{r}
write.csv(res,file = "myresults.csv")
```

```{r}
save(res,file="my_results.RData")
```
